geode-core/src/main/java/org/apache/geode/internal/cache/TXReservationMgr.java

```diff
@@ -96,15 +84,19 @@ private void checkForConflict(TXRegionLockRequestImpl rr, IdentityArrayList loca
     }
   }
 
-  private void checkSetForConflict(TXRegionLockRequestImpl rr, Set s, Object[] keys,
-      IdentityArrayList localLocks) throws CommitConflictException {
-    for (int i = 0; i < keys.length; i++) {
-      if (s.contains(keys[i])) {
-        release(localLocks, true);
-        throw new CommitConflictException(
-            String.format(
-                "The key %s in region %s was being modified by another transaction locally.",
-                new Object[] {keys[i], rr.getRegionFullPath()}));
+  private void checkSetForConflict(TXRegionLockRequestImpl rr, Map<Object, Boolean> oldValue,
+      Map<Object, Boolean> keys, IdentityArrayList localLocks) throws CommitConflictException {
+    for (Map.Entry<Object, Boolean> e : keys.entrySet()) {
+      if (oldValue.containsKey(e.getKey())) {
+        if (oldValue.get(e.getKey())) {
+          release(localLocks, true);
+          throw new CommitConflictException(
+              String.format(
+                  "The key %s in region %s was being modified by another transaction locally.",
+                  new Object[] {e.getKey(), rr.getRegionFullPath()}));
+        } else {
+          oldValue.remove(e.getKey());
```

Comment:
oldValue should not be changed. This will be used to restore existing locks held by other transactions if the current transaction detects a conflict and fails.